rust/src/sdp/parser.rs

```diff
@@ -0,0 +1,694 @@
+/* Copyright (C) 2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// written by Giuseppe Longo <giuseppe@glongo.it>
+
+use nom7::{
+    branch::alt,
+    bytes::complete::{tag, take_till, take_while, take_while_m_n},
+    character::{
+        complete::{char as char_parser, digit1, line_ending, space1, u8 as take_u8},
+        is_alphabetic,
+    },
+    character::{is_alphanumeric, is_digit, is_space},
+    combinator::map_res,
+    combinator::{opt, peek},
+    error::{make_error, ErrorKind},
+    multi::{many0, many1},
+    number::complete::be_u8,
+    sequence::{preceded, tuple},
+    {Err, IResult},
+};
+
+use std::net::IpAddr;
+use std::str::FromStr;
+
+#[derive(Debug)]
+pub struct SdpMessage {
+    pub version: u32,
+    pub origin: OriginField,
+    pub session_name: String,
+    pub session_info: Option<String>,
+    pub uri: Option<String>,
+    pub email: Option<String>,
+    pub phone_number: Option<String>,
+    pub connection_data: Option<ConnectionData>,
+    pub bandwidths: Option<Vec<String>>,
+    pub time: String,
+    pub repeat_time: Option<String>,
+    pub time_zone: Option<String>,
+    pub encryption_key: Option<String>,
+    pub attributes: Option<Vec<String>>,
```

Comment:
Could we just have `Vec` without `Option` and keep it empty when there is nothing ?

Reply:
No, because `attributes` are provided by the `parse_attributes` parser, which is applied `opt`ionally.

Reply:
But do we need to call `parse_attributes` with `opt` ? Would it not return an empty Vec successfully instead ?

Reply:
Yes, because `attributes` are optional and not mandatory.